tron/mesos.py

```diff
@@ -0,0 +1,272 @@
+                self.log.error('Unknown failure, exiting')
+
+            self.done()
+
+
+class MesosCluster:
+    def __init__(self, mesos_address):
+        self.mesos_address = mesos_address
+        self.processor = get_mesos_processor()
+        self.queue = PyDeferredQueue()
+        self.deferred = None
+        self.runner = None
+        self.tasks = {}
+
+        self.connect()
+
+    # TODO: Should this be done asynchronously?
+    # TODO: Handle/retry errors
+    def connect(self):
+        self.runner = self.get_runner(self.mesos_address, self.queue)
+        self.handle_next_event()
+
+    def handle_next_event(self, deferred_result=None):
+        if self.deferred and not self.deferred.called:
+            log.warn(
+                'Already have handlers waiting for next event in queue, '
+                'not adding more'
+            )
+            return
+        self.deferred = self.queue.get()
+        self.deferred.addCallback(self._process_event)
+        self.deferred.addCallback(self.handle_next_event)
+        self.deferred.addErrback(logError)
+        self.deferred.addErrback(self.handle_next_event)
+
+    def submit(self, task):
+        if self.runner.stopping:
+            # Last framework was terminated for some reason, re-connect.
+            self.connect()
+        elif self.deferred.called:
+            # Just in case callbacks are missing, re-add.
+            self.handle_next_event()
+
+        mesos_task_id = task.get_mesos_id()
+        self.tasks[mesos_task_id] = task
+        self.runner.run(task.get_config())
+        log.info(
+            'Submitting task {} to {}'.format(
+                mesos_task_id,
+                self.mesos_address,
+            ),
+        )
+
+    def create_task(
+        self,
+        action_run_id,
+        command,
+        cpus,
+        mem,
+        constraints,
+        docker_image,
+        docker_parameters,
+        env,
+        extra_volumes,
+        serializer,
+    ):
+        task_config = self.runner.TASK_CONFIG_INTERFACE(
+            name=action_run_id,
+            cmd=command,
+            cpus=cpus,
+            mem=mem,
+            constraints=constraints,
+            image=docker_image,
+            docker_parameters=docker_parameters,
+            environment=env,
+            volumes=extra_volumes,  # TODO: add default volumes
+            uris=[DOCKERCFG_LOCATION],
+            offer_timeout=OFFER_TIMEOUT,
+        )
+        return MesosTask(action_run_id, task_config, serializer)
+
+    def get_runner(self, mesos_address, queue):
+        executor = self.processor.executor_from_config(
+            provider='mesos',
+            provider_config={
+                'secret': MESOS_SECRET,
+                'mesos_address': get_mesos_leader(mesos_address),
+                'role': MESOS_ROLE,
+                # TODO: could also be in config, to include Tron cluster name
+                'framework_name': 'tron',
```

Comment:
should this be parameterised if we're leaving the possibility of >1 framework open (the frameworks object in this module is a dict, rather than a single object).

Reply:
There's only supposed to be one framework per cluster, so it shouldn't register the same framework name on the same cluster more than once. But adding the parameter seems like a good idea anyway, so I'll do that.

Reply:
Er, just realized that we don't really have the Mesos cluster name, just the fqdn. Is there a different parameter you think would be useful?

Reply:
how'd you mean? this is the gramework name, rather than the cluster name?